### PR TITLE
Resource Bundle Look-ups in Modular Java Applications

### PR DESCRIPTION
**Pull request by gunnarmorling**

### CODE REVIEW

File: content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+But unfortunately, extensions aren't evaluated by the look-up routine in `ResourceBundle`.
+So instead we'll go with propagating the key namespace information via the locale's variant.
+First, let's revisit the code in the `ErrorHandler` class:
+
+[source,java,linenums=true]
+----
+public class ErrorHandler {
+
+  public String getErrorMessage(String key, UserContext context) {
+      String component = key.split("\\.")[0]; // <1>
```
Comment (hpgrahsl) [inline content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc:156]: nit pick -> why is the var named 'component'. if I didn't miss it the text never talks about components but bundles, modules and key prefixes?

File: content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+      </manifestEntries>
+    </archive>
+  </configuration>
+</plugin>
+...
+----
+<1> Set up another execution of the https://maven.apache.org/plugins/maven-compiler-plugin/[Maven compiler plug-in] for the Java 9 specific sources,
+<2> using Java 9 bytecode level,
+<3> picking up the sources from _src/main/java-9_,
+<4> and organizing the compilation output in the multi-relese structure under _META-INF/versions/..._
```
Comment (hpgrahsl) [inline content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc:382]: multi-rele*a*se

**Review by hpgrahsl — COMMENTED** [review]
👏 Very nice. Learned new stuff by reading this! 🙇‍♂️ for sharing.

File: content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+module dev.morling.links.core {
+  exports dev.morling.links.core;
+  exports dev.morling.links.core.spi; // <1>
+  uses dev.morling.links.core.spi.LinksMessagesProvider; // <2>
+}
+----
+<1> Export the package of the resource bundle provider interface so that implementations can be created in other modules
+<2> Declare the usage of the `LinksMessagesProvider` service
+
+Using the resource bundle in the error handler class is rather un-exciting;
```
Comment (gunnarmorling) [inline content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc:97]:
```suggestion
Using the resource bundle in the error handler class is rather unexciting;
```

[the remark this answers is not in the document]
Comment (gunnarmorling) [reply content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc:97]: Thx, changed.

File: content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+The idea to address this concern is the following:
+
+* Prefix each message key with a module specific string, resulting in keys like `tournament.fullybooked`, `greenkeeping.greenclosed`, etc.
+* When requesting the bundle for a given key in the error handler class, obtain the key's prefix and pass it to bundle providers
+* Let bundle providers react only to their specific message prefix
+
+This is where things become a little bit fiddly:
+there isn't a really good way for passing such contextual information from bundle consumers to providers.
+Our loop hole here will be to squeeze that information into the the requested https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Locale.html[`Locale`] instance.
+Besides the well-known language and country attributes, `Locale` can also carry _variant_ data and even application specific https://docs.oracle.com/javase/tutorial/i18n/locale/extensions.html[_extensions_].
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc:144]: Yes, that'd be the idea. If it worked :) I'll clarify it.

File: content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc (hunk 1 above)
Comment (gunnarmorling) [reply content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc:156]: Good catch, I was/am using the terms component and modules somewhat interchangeable. I'll adjust it.

File: content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+----
+public class ErrorHandler {
+
+  public String getErrorMessage(String key, UserContext context) {
+      String component = key.split("\\.")[0]; // <1>
+      Locale locale = new Locale( // <2>
+          context.getLocale().getLanguage(),
+          context.getLocale().getCountry(),
+          component
+      );
```
Comment (gunnarmorling) [inline content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc:161]:
```suggestion
  public String getErrorMessage(String key, UserContext context) {
      String prefix = key.split("\\.")[0]; // <1>
      Locale locale = new Locale( // <2>
          context.getLocale().getLanguage(),
          context.getLocale().getCountry(),
          prefix
      );
```

File: content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc (hunk 2 above)
Comment (gunnarmorling) [inline content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc:382]:
```suggestion
<4> and organizing the compilation output in the multi-release structure under _META-INF/versions/..._
```

File: content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+
+With a little bit of extra effort, it also was possible to design the code in the _core_ module in a way that the application can be used with different Java versions and configurations:
+on the module path with Java 9+, on the classpath with Java 9+, on the classpath with Java 8.
+
+If you'd like to explore the complete code by yourself, you can find it in the https://github.com/gunnarmorling/modular-resource-bundles[modular-resource-bundles] GitHub repository.
+To learn more about resource bundle retrieval in named modules,
+please refer to the extensive documentation of https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ResourceBundle.html#resource-bundle-modules[`ResourceBundle`] and https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/spi/ResourceBundleProvider.html[`ResourceBundleProvider`].
```
Comment (gunnarmorling) [inline content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc:399]:
```suggestion
please refer to the extensive documentation of https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ResourceBundle.html#resource-bundle-modules[`ResourceBundle`] and https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/spi/ResourceBundleProvider.html[`ResourceBundleProvider`].

_Many thanks to https://twitter.com/hpgrahsl[Hans-Peter Grahsl] for providing feedback while writing this post!_
```

**Review by gunnarmorling — COMMENTED** [review]
Thx for reviewing, @hpgrahsl! Will push a commit to address your remarks in a minute.

**Pull request merged**